### DISCUSSION TITLE
Use runno-clang to compile cpp code locally

### DISCUSSION THREAD

**Discussion by swwind** [taybenlor/runno#213] [Q&A]
Hello @taybenlor, I'm trying to implement a little runno-like plugin for my project nowadays, but I'm having a headache trying to solve some complicated issues with `clang` and `runno-clang`.

I have my machine with wasmer 2.3.0 and wapm 0.5.5 installed, and I'm trying to compile my cpp code with `runno-clang`. I copied compile command from `provider.ts` and try to run it on my machine, but it fails.

```cpp
// example.cpp
#include <stdio.h>
int main() {
  printf("hello world");
}
```

```
wapm run runno-clang --dir=. \
  -cc1 \
  -Werror \
  -emit-obj \
  -disable-free \
  -isysroot /sys \
  -internal-isystem /sys/include/c++/v1 \
  -internal-isystem /sys/include \
  -internal-isystem /sys/lib/clang/8.0.1/include \
  -ferror-limit 4 \
  -fmessage-length 80 \
  -fcolor-diagnostics \
  -O2 \
  -o program.o \
  -x c++ \
  -v \
  ./example.cpp
```

```
clang -cc1 version 8.0.1 based upon LLVM 8.0.1 default target wasm32-wasi
ignoring nonexistent directory "/sys/usr/local/include"
ignoring nonexistent directory "../lib/clang/8.0.1/include"
ignoring nonexistent directory "/sys/usr/include"
ignoring duplicate directory "/sys/include/c++/v1"
ignoring duplicate directory "/sys/include/c++/v1"
#include "..." search starts here:
#include <...> search starts here:
 /sys/include/c++/v1
End of search list.
In file included from ./example.cpp:1:
/sys/include/c++/v1/stdio.h:108:15: fatal error: 'stdio.h' file not found
#include_next <stdio.h>
              ^~~~~~~~~
1 error generated.
```

The weird is that two `ignoring duplicate directory "/sys/include/c++/v1"` errors in clang, I have three different include paths but it seems only receives the first include path and ignores others. I'm not sure who causes the problem but the same command and code seems to work in runno.run.

P.S. `clang` also have the same behavior (ignores other include paths).

Do you have any idea on this?

**Comment by swwind** [comment]
![image](https://user-images.githubusercontent.com/26316494/196171822-721a0c12-e5a1-4163-b0e2-306d18341052.png)

And runno.run works very well :cry:

**Comment by taybenlor** [comment]
Hi! Thanks for reaching out!

Not sure how helpful I can be. I think the important output here is:

```
ignoring nonexistent directory "/sys/usr/local/include"
ignoring nonexistent directory "../lib/clang/8.0.1/include"
ignoring nonexistent directory "/sys/usr/include"
ignoring duplicate directory "/sys/include/c++/v1"
ignoring duplicate directory "/sys/include/c++/v1"
```

If you have a look at https://github.com/taybenlor/clang you'll see that it includes a `sysroot` folder. If you look at the `wapm.toml` file it sets up the `sysroot` to be mounted to `/sys`. I'm guessing this step isn't being done automatically by WAPM for some reason?

BTW my https://github.com/taybenlor/clang (`runno-clang` on WAPM) is just a fork of `clang` on WAPM. I just replaced the `wasm` binaries with the original ones from binji's talk which seemed to work better for C++.

**Reply by swwind** [reply]
As a matter of fact, `wapm` does the mount job automatically, it more looks like the problem to `clang` or the `wasmer` runner itself. The thing is, `clang` does not respect my `-internal-isysroot` argument and thus cannot find where the include file located.

I found a similar crash report [here](https://github.com/taybenlor/runno/pull/195#pullrequestreview-937074019), the same compile command does not seem to work on native environments.

I may report this to `wasmer` if this does not seem to be the `clang.wasm` binary problem.

**Reply by swwind** [reply]
If you give 1 `-internal-isystem /sys/include/c++/v1` and 1000 `-internal-isystem /sys/include`, you will get 1000 `ignoring duplicate directory "/sys/include/c++/v1"`. There was something definitely wrong.

**Reply by taybenlor** [reply]
How'd you go with this? Need any further help?

**Reply by swwind** [reply]
I found it was a `wasmer` bug, and `wasmtime` works. So I decided to use `wasmtime` instead.

**Reply by taybenlor** [reply]
Great, thanks!
